Cannot build with IPASIR - cryptominisat

<!---
Thank you for reporting a problem and suggesting improvements. Please provide the below information to make sure we can effectively deal with the issue reported. For the most precise version information, see the first line of console output or run with --version. Please attach or include example code that allows us to reproduce the problem.
--->

CBMC version: master and Release 5.12
Operating system: Linux (Debian and Ubuntu)
Exact command line resulting in the issue: make -C src IPASIR=/home/ubuntu/cryptominisat/src LIBS="/home/ubuntu/cryptominisat/build/lib/libipasircryptominisat5.a"
(cryptominisat built OK)
What behaviour did you expect: Successful build
What happened instead:
# Entering solvers
make  -C solvers
make[1]: Entering directory '/home/ubuntu/cbmc-cbmc-5.12/src/solvers'
g++ -c -DHAVE_IPASIR -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MMD -MP -std=c++11 -O2 -DHAVE_IPASIR   -I ..     -I /home/ubuntu/cryptominisat/src       -o sat/satcheck_ipasir.o sat/satcheck_ipasir.cpp
sat/satcheck_ipasir.cpp: In constructor ‘satcheck_ipasirt::satcheck_ipasirt()’:
sat/satcheck_ipasir.cpp:161:36: error: no matching function for call to ‘cnf_solvert::cnf_solvert()’
 satcheck_ipasirt::satcheck_ipasirt()
                                    ^
In file included from sat/satcheck_ipasir.h:15:0,
                 from sat/satcheck_ipasir.cpp:20:
sat/cnf.h:72:12: note: candidate: cnf_solvert::cnf_solvert(message_handlert&)
   explicit cnf_solvert(message_handlert &message_handler)
            ^~~~~~~~~~~
sat/cnf.h:72:12: note:   candidate expects 1 argument, 0 provided
sat/cnf.h:69:7: note: candidate: cnf_solvert::cnf_solvert(const cnf_solvert&)
 class cnf_solvert:public cnft
       ^~~~~~~~~~~
sat/cnf.h:69:7: note:   candidate expects 1 argument, 0 provided
sat/cnf.h:69:7: note: candidate: cnf_solvert::cnf_solvert(cnf_solvert&&)
sat/cnf.h:69:7: note:   candidate expects 1 argument, 0 provided
../common:222: recipe for target 'sat/satcheck_ipasir.o' failed
make[1]: *** [sat/satcheck_ipasir.o] Error 1
make[1]: Leaving directory '/home/ubuntu/cbmc-cbmc-5.12/src/solvers'
Makefile:97: recipe for target 'solvers.dir' failed
make: *** [solvers.dir] Error 2
make: Leaving directory '/home/ubuntu/cbmc-cbmc-5.12/src'

--- a/src/Makefile
+++ b/src/Makefile
@@ -163,4 +163,4 @@ cadical-download:
 doc :
 	doxygen
 
-.PHONY: ipasir-build minisat2-download glucose-download
+.PHONY: minisat2-download cudd-download glucose-download cadical-download

--- a/src/memory-analyzer/Makefile
+++ b/src/memory-analyzer/Makefile
@@ -7,7 +7,7 @@ SRC = \
 
 INCLUDES= -I ..
 
-LIBS = \
+OBJ += \
   ../ansi-c/ansi-c$(LIBEXT) \
   ../goto-programs/goto-programs$(LIBEXT) \
   ../linking/linking$(LIBEXT) \
@@ -15,6 +15,7 @@ LIBS = \
   ../big-int/big-int$(LIBEXT) \
   ../langapi/langapi$(LIBEXT)
 
+LIBS =
 
 CLEANFILES = memory-analyzer$(EXEEXT)
 

--- a/src/solvers/sat/satcheck_ipasir.cpp
+++ b/src/solvers/sat/satcheck_ipasir.cpp
@@ -90,6 +90,9 @@ void satcheck_ipasirt::lcnf(const bvt &bv)
   }
   ipasir_add(solver, 0); // terminate clause
 
+  with_solver_hardness(
+    [&bv](solver_hardnesst &hardness) { hardness.register_clause(bv); });
+
   clause_counter++;
 }
 
@@ -158,8 +161,8 @@ void satcheck_ipasirt::set_assignment(literalt a, bool value)
   INVARIANT(false, "method not supported");
 }
 
-satcheck_ipasirt::satcheck_ipasirt()
-: solver(nullptr)
+satcheck_ipasirt::satcheck_ipasirt(message_handlert &message_handler)
+  : cnf_solvert(message_handler), solver(nullptr)
 {
   INVARIANT(!solver, "there cannot be a solver already");
   solver=ipasir_init();

--- a/src/solvers/sat/satcheck_ipasir.h
+++ b/src/solvers/sat/satcheck_ipasir.h
@@ -14,11 +14,13 @@ instructions.
 
 #include "cnf.h"
 
+#include <solvers/hardness_collector.h>
+
 /// Interface for generic SAT solver interface IPASIR
-class satcheck_ipasirt:public cnf_solvert
+class satcheck_ipasirt : public cnf_solvert, public hardness_collectort
 {
 public:
-  satcheck_ipasirt();
+  satcheck_ipasirt(message_handlert &message_handler);
   virtual ~satcheck_ipasirt() override;
 
   /// This method returns the description produced by the linked SAT solver
@@ -44,12 +46,28 @@ public:
     return true;
   }
 
+  void
+  with_solver_hardness(std::function<void(solver_hardnesst &)> handler) override
+  {
+    if(solver_hardness.has_value())
+    {
+      handler(solver_hardness.value());
+    }
+  }
+
+  void enable_hardness_collection() override
+  {
+    solver_hardness = solver_hardnesst{};
+  }
+
 protected:
   resultt do_prop_solve() override;
 
   void *solver;
 
   bvt assumptions;
+
+  optionalt<solver_hardnesst> solver_hardness;
 };
 
 #endif // CPROVER_SOLVERS_SAT_SATCHECK_IPASIR_H

